src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -739,11 +787,87 @@ where
+        SourceReaderStreams {
+            batches: batch_stream,
+            batch_upper_summaries: summary_stream,
+            resumption_feedback_stream: feedback_stream,
+            health_stream,
+        },
         Some(capability),
     )
 }
 
+/// Mints new contents for the remap shard based on summaries about the source
+/// upper it receives from the raw reader operators.
+///
+/// Only one worker will be active and write to the remap shard. All source
+/// upper summaries will be exchanged to it.
+fn health_operator<G>(
+    scope: &G,
+    config: RawSourceCreationConfig,
+    health_stream: timely::dataflow::Stream<G, (usize, HealthStatus)>,
+) -> Rc<dyn Any>
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let RawSourceCreationConfig {
+        worker_id,
+        worker_count,
+        id: source_id,
+        ..
+    } = config;
+
+    // We'll route all the work to a single arbitrary worker;
+    // there's not much to do, and we need a global view.
+    let chosen_worker_id = 0;
+
+    let mut healths = vec![HealthStatus::Starting; worker_count];
+
+    let operator_name = format!("healthcheck({})", worker_id);
+    let mut health_op = OperatorBuilder::new(operator_name, scope.clone());
+
+    let mut input = health_op.new_input_connection(
+        &health_stream,
+        Exchange::new(move |_| chosen_worker_id as u64),
+        vec![],
+    );
+
+    fn overall_status(healths: &[HealthStatus]) -> &HealthStatus {
+        healths.iter().max().unwrap_or(&HealthStatus::Starting)
+    }
+
+    let mut last_reported_status = overall_status(&healths).clone();
+
+    health_op.build_async(
+        scope.clone(),
+        move |mut _capabilities, _frontiers, scheduler| async move {
+            let mut buffer = Vec::new();
+            info!("Health for source {source_id} initialized to: {last_reported_status:?}");
+            while scheduler.notified().await {
+                input.for_each(|_cap, rows| {
+                    rows.swap(&mut buffer);
+                    for (worker_id, health_event) in buffer.drain(..) {
+                        if worker_id != chosen_worker_id {
+                            warn!("Health messages for source {source_id} passed to an unexpected worker id: {worker_id}")
+                        }
+                        let prev_health = &healths[worker_id];
+                        if prev_health != &health_event {
+                            healths[worker_id] = health_event;
+                            let new_status = overall_status(&healths);
+                            if &last_reported_status != new_status {
+                                info!("Health transition for source {source_id}: {last_reported_status:?} -> {new_status:?}");
+                                last_reported_status = new_status.clone();
+                            }
+                        }
+                    }
+                })
+            }
+        },
+    );
+
+    Rc::new(())
```

Comment:
should we do something like the remap operator w.r.t. the token?

That operator (and a few others) creates a token and a weak reference -- and on each invocation checks if the weak token can be upgraded.  If not, it drops the operators:
```
    let token = Rc::new(());
    let token_weak = Rc::downgrade(&token);
[...]
            while scheduler.notified().await {
                if token_weak.upgrade().is_none() {
                    // Make sure we don't accidentally mint new updates when
                    // this source has been dropped. This way, we also make sure
                    // to not react to spurious frontier advancements to `[]`
                    // that happen when the input source operator is shutting
                    // down.
                    return;
                }
 ```

If the token is just an `Rc` that we don't use anywhere, then we don't need to return a token at all.  The tokens are generally used to communicate to the operators when they should shut down -- so if we don't need that information, then there's no use in having a token!

That being said, I _think_ we do want a token so we can early shutdown like most of the other `build_async` ops.  But you've thought about this more than me: any reason we don't actually need that functionality?

Reply:
Hmm seems likely! What's expected to happen if we don't?

Reply:
Discussed elsewhere. The takeaway was that, while this is probably not necessary for the semantics of this particular thing, it's best to stay consistent with the overall source for now. If we want to use the more "timely native" shutdown semantics we can do that later.